### ISSUE TITLE
Add multi-threading support to Lua

### ISSUE THREAD

**Issue by MegadreamsBE** [multitheftauto/mtasa-blue#918]
**Is your feature request related to a problem? Please describe.**

While Lua is fast we are still bound to limitations that can make it hard to do some expensive calculations for things like path finding or dealing with running a SCM virtual machine (Sphene). It would be much nicer if this code could be moved into a separate thread and therefore prevent freezes.

Lua itself doesn't support multi-threading but is has been shown [here](https://love2d.org/wiki/love.thread) that there is definitely a way to do so.

**Describe the solution you'd like**

I'd like for new functions to be exposed that allow for a thread to be created and for data to be transmitted between them. In line with the example of LÖVE I propose the following functions to be added at least:

- createThread(filePath) ( **OOP:** Thread(filePath) )
Creates a new thread and returns the newly created thread.
_Parameters:_ 
  - file: Path to a file to be ran in the thread or a string containing Lua code.
- killThread(thread) ( **OOP:** thread:kill() )
Terminates a thread forcefully, ending its execution.
_Parameters:_ 
  - thread: The thread to be terminated.
- startThread(thread) ( **OOP:** thread:start() )
Starts the thread.
_Parameters:_ 
  - thread: The thread to be started.
- isThreadRunning(thread) ( **OOP:** thread:isRunning() )
Returns true if the thread is running and false if not.
_Parameters:_ 
  - thread: The thread to be checked.
- sendThreadMessage(thread, value) ( **OOP:** thread:sendMessage(value) )
Sends a message to the thread.
_Parameters:_ 
  - thread: The thread to send the message to.
  - value: The value to send to the thread.
- receiveThreadMessage() ( **OOP:** Thread.receiveMessage() )
Receives a message from the message queue for the thread in which this is executed. Although this could also be a event (perhaps both?). Returns the thread that has sends the message and the value of that message.
- getThreadMessageCount() ( **OOP:** Thread.getMessageCount() )
Returns the amount of messages in the message queue for the thread in which this is executed.

Please note that this is an initial thought and that this API should be heavily discussed and adjusted where possible.

**Describe alternatives you've considered**

Using coroutines but you can only run one at once so it can slow down execution even more if improperly used. Coroutines work better for things like task scheduling.

**Comment by qaisjp** [comment]
I like love's approach with Channels. Very powerful imo 

There was some discussion about this on #mta.dev (to be copied here)

**Comment by MegadreamsBE** [comment]
I agree, wasn't sure if it would have been easily accepted here so I went for a more simple API but figured it would come up if desired. Glad it did. Do you propose to keep a similar API for the channels as LÖVE or with modifications to it?

**Comment by qaisjp** [comment]
I would be happy with an identical or similar API.

**Comment by sbx320** [comment]
I'd love to get `createThread(function() return 3 end)` working. Not quite sure how though. Would need some investigation.

**Comment by MegadreamsBE** [comment]
That would be ideal and from what I can read online possible but it's rather delicate to move a function to a different Lua state so it will require a lot of testing to make sure the garbage collector for a start doesn't get confused..

**Comment by qaisjp** [comment]
> I'd love to get `createThread(function() return 3 end)` working. Not quite sure how though. Would need some investigation.

`string.dump` is usable in the simplest case

(upvalues would not be copied, to keep behaviour between `createThread(filename)` and `createThread(fn)` consistent ... but this means you can't call other "non-stdlib" functions)

**Comment by jushar** [comment]
I stumbled upon this project (https://github.com/effil/effil) today. Perhaps we can use that or at least consider it as an inspiration.

**Comment by qaisjp** [comment]
> ## [`runner = effil.thread(func)`](https://github.com/effil/effil#runner--effilthreadfunc)
> 
> Creates thread runner. Runner spawns new thread for each invocation.
>
> **input**: _func_ - Lua function

@sbx320 I remember you were saying that you really wanted something like this where you could just pass a function

Aha:

> Functions are dumped using `lua_dump`. Upvalues are captured according to [the rules](https://github.com/effil/effil#functions-upvalues).

**Comment by MegadreamsBE** [comment]
Since over at the Sphene project we have a big need for threading, I guess I'll give implementing it a try myself. Would highly appreciate any help of course.
